codegen/src/main/java/software/amazon/awssdk/codegen/poet/auth/scheme/AuthSchemeParamsSpec.java

```diff
@@ -120,11 +135,53 @@ private void addBuilderSetterMethods(TypeSpec.Builder b) {
         if (authSchemeSpecUtils.usesSigV4()) {
             b.addMethod(MethodSpec.methodBuilder("region")
                                   .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
+                                  // TODO: Use Region.class instead?
                                   .addParameter(ParameterSpec.builder(String.class, "region").build())
                                   .returns(authSchemeSpecUtils.parametersInterfaceBuilderInterfaceName())
                                   .addJavadoc("Set the region. The region parameter may be used with  the $S auth scheme.",
                                               "aws.auth#sigv4") // TODO: Reference the SigV4 AuthScheme when implemented
                                   .build());
         }
+
+        if (authSchemeSpecUtils.generateEndpointBasedParams()) {
+            parameters().forEach((name, model) -> {
+                String methodName = authSchemeSpecUtils.paramMethodName(name);
+                if (!"region".equals(methodName) || !authSchemeSpecUtils.usesSigV4()) {
+                    b.addMethod(setterMethodDeclaration(name, model));
+                }
+            });
+        }
+    }
+
+    private MethodSpec setterMethodDeclaration(String name, ParameterModel model) {
+        MethodSpec.Builder spec = paramMethodBuilder(name, model)
+            .addModifiers(Modifier.ABSTRACT)
+            .addParameter(parameterSpec(name, model))
+            .returns(authSchemeSpecUtils.parametersInterfaceBuilderInterfaceName());
+        if (model.getDocumentation() != null) {
+            spec.addJavadoc(model.getDocumentation());
+        }
+        return spec.build();
+    }
+
+    private ParameterSpec parameterSpec(String name, ParameterModel model) {
+        return ParameterSpec.builder(parameterType(model), variableName(name)).build();
+    }
+
+    private String variableName(String name) {
+        return intermediateModel.getNamingStrategy().getVariableName(name);
+    }
```

Comment:
seems like instead `parameterSpec` could be static methods in EndpointRulesSpecUtils.